_posts/2020-10-27-a-closer-look-at-the-interpreter.md

```diff
@@ -0,0 +1,202 @@
+}
+```
+
+There's a bit more to it, and those who would like to know more can read the
+[documentation for the `beam_makeops` script], but the above covers the gist of
+it.
+
+While the above is quite efficient, there's significant overhead from dispatch
+and argument handling. Here's a slightly altered assembly listing of the
+above:
+
+<pre class="highlight">
+    <em>; Load the argument word and extract the offset of our
+    ; source register, which lives in the upper 32 bits of
+    ; said word.</em>
+    mov    rdx, [rbx + 8]
+    mov    rcx, rdx
+    shr    rcx, 32
+
+    <em>; Our source register is at address [r15 + rcx]. Check
+    ; whether it contains a list. </em>
+    <b>test   byte [r15 + rcx], (_TAG_PRIMARY_MASK - TAG_PRIMARY_LIST)
+    jne    jump_to_fail_label</b>
+
+    <em>; Jump to the instruction directly following this one.</em>
+    add    rbx, 16
+    jmp    [rbx]
+
+jump_to_fail_label:
+    <em>; Extract the offset of our fail label from the
+    ; argument word. It lives in the lower 32 bits so
+    ; we'll sign-extend it to the full 64 bits.</em>
+    movsxd rdx, edx
+
+    <em>; Add our offset to the instruction pointer and then
+    ; jump there.</em>
+    lea    rbx, [rbx + rdx * 8]
+    jmp    [rbx]
+</pre>
+
+The bold section is the meat of the instruction and the rest is argument
+unpacking and instruction dispatch. While this is not much of a problem for
+large instructions, its effect on short ones like this is very large, and when
+looking through a profiler (e.g. [`perf`]) it's not unusual for the final `jmp`
+to dominate the rest.
+
+To lessen this effect, the loader combines commonly used instruction sequences
+into a single instruction. For example, two independent moves may fuse into
+[`move2_par`], and `is_nonempty_list` followed by `get_list` might be fused to
+[`is_nonempty_list_get_list`].
+
+This reduces the cost of short instructions, but only works to the extent we're
+able to identify common patterns and comes at a significant maintenance cost
+as each combination must be implemented manually. Even so, the effect tends to
+be moderate and the dispatch overhead remains significant.
+
+Another, albeit lesser, downside with the interpreter is that modern processors
+are very optimized for patterns commonly found in "ordinary" native code. For
+example, nearly all of them have a special branch predictor just for calls and
+returns. Assuming that every call has a corresponding return lets it predict
+returns _perfectly_ unless an exception is thrown, but since the interpreter
+does not use native calls and returns, it cannot make use of this optimization.
+
+Unfortunately there's not a whole lot that can be done about this, and after
+over two decades of refinement it's becoming increasingly difficult to optimize
+it in meaningful ways.
+
+Because of that our quest to improve performance has instead focused on two
+areas; improving the compiler, and implementing a JIT. We've made great strides
```

Comment:
Semicolon is not "super colon". The [Oxford comma](https://en.wikipedia.org/wiki/Serial_comma) should only be used when there is a series of three or more items.

```suggestion
areas: improving the compiler and implementing a JIT. We've made great strides
```